### PR TITLE
Block SyncSets with secret ref in another namespace.

### PR DESCRIPTION
**Pull request by dgoodwin**
Blocked during admission, but as a secondary check we also make sure not
to reconcile such a SyncSet and log a warning.

This change is not relevant for SelectorSyncSets which are cluster
scoped.

### CODE REVIEW

File: pkg/apis/hive/v1/validating-webhooks/syncset_validating_admission_hook.go [hunk 1]
```diff
@@ -311,6 +313,19 @@ func validateSecrets(secrets []hivev1.SecretMapping, fldPath *field.Path) field.
 	return allErrs
 }
 
+func validateSourceSecretInSyncSetNamespace(secrets []hivev1.SecretMapping, syncSetNS string, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	for i, secret := range secrets {
+		if secret.SourceRef.Namespace != syncSetNS {
```
Comment (staebler) [inline pkg/apis/hive/v1/validating-webhooks/syncset_validating_admission_hook.go:319]: This should also allow an empty namespace, which implies the same namespace as the syncset.

File: pkg/controller/syncsetinstance/syncsetinstance_controller.go [hunk 2]
```diff
@@ -264,6 +264,15 @@ func (r *ReconcileSyncSetInstance) Reconcile(request reconcile.Request) (reconci
 		return reconcile.Result{}, err
 	}
 
+	// Prevent users with permissions to create a SyncSet from referencing Secrets in another namespace
+	// and copying them into their clusters. This should be caught in admission, but just incase we have
```
Comment (staebler) [inline pkg/controller/syncsetinstance/syncsetinstance_controller.go:268]: s/incase/in case/

File: pkg/controller/syncsetinstance/syncsetinstance_controller.go [hunk 3]
```diff
@@ -303,6 +312,19 @@ func (r *ReconcileSyncSetInstance) Reconcile(request reconcile.Request) (reconci
 	return reconcile.Result{RequeueAfter: reapplyDuration}, nil
 }
 
+func (r *ReconcileSyncSetInstance) hasSourceSecretsInAnotherNamespace(ssi *hivev1.SyncSetInstance, spec *hivev1.SyncSetCommonSpec) bool {
+	// This check is not relevant for SelectorSyncSets
+	if ssi.Spec.SyncSetRef == nil {
```
Comment (staebler) [inline pkg/controller/syncsetinstance/syncsetinstance_controller.go:317]: Do we need this check? This was already checked by the caller. Can we either remove it here or remove it at the caller location?

File: pkg/controller/syncsetinstance/syncsetinstance_controller.go [hunk 4]
```diff
@@ -303,6 +312,19 @@ func (r *ReconcileSyncSetInstance) Reconcile(request reconcile.Request) (reconci
...
+		return false
+	}
+	for _, secret := range spec.Secrets {
+		if secret.SourceRef.Namespace != ssi.Namespace {
```
Comment (staebler) [inline pkg/controller/syncsetinstance/syncsetinstance_controller.go:321]: Allow empty namespace.

File: pkg/controller/syncsetinstance/syncsetinstance_controller.go [hunk 5]
```diff
@@ -264,6 +264,15 @@ func (r *ReconcileSyncSetInstance) Reconcile(request reconcile.Request) (reconci
 		return reconcile.Result{}, err
 	}
 
+	// Prevent users with permissions to create a SyncSet from referencing Secrets in another namespace
+	// and copying them into their clusters. This should be caught in admission, but just incase we have
+	// a check here to refuse to reconcile. Error is logged but not returned as there's no point re-reconciling
+	// this unless the SyncSet is edited. This check is not relevant for SelectorSyncSets which have no namespace.
+	if ssi.Spec.SyncSetRef != nil && r.hasSourceSecretsInAnotherNamespace(ssi, spec) {
+		ssiLog.Warn("syncsetinstance has a source secret in another namespace")
```
Comment (staebler) [inline pkg/controller/syncsetinstance/syncsetinstance_controller.go:272]: In case this does get through admission validation, this should bubble up as a condition on the SyncSetInstance. Otherwise, the user will not know why the SyncSet is not applied.

Comment (dgoodwin) [reply pkg/controller/syncsetinstance/syncsetinstance_controller.go:272]: Because this is supposed to be handled elsewhere and is just a backup check, it felt like we could get away without carrying code to set a condition similar to some of the other checks above this. Would you agree?

Comment (staebler) [reply pkg/controller/syncsetinstance/syncsetinstance_controller.go:272]: I can live with that. Do we know if there are existing syncsets that violate this new validation?

File: pkg/apis/hive/v1/validating-webhooks/syncset_validating_admission_hook.go (hunk 1 above)
Comment (dgoodwin) [reply pkg/apis/hive/v1/validating-webhooks/syncset_validating_admission_hook.go:319]: Done and added unit test.

File: pkg/controller/syncsetinstance/syncsetinstance_controller.go (hunk 5 above)
Comment (dgoodwin) [reply pkg/controller/syncsetinstance/syncsetinstance_controller.go:272]: Scanned stage, doesn't look like there are any violating.

File: pkg/controller/syncsetinstance/syncsetinstance_controller.go (hunk 4 above)
Comment (dgoodwin) [reply pkg/controller/syncsetinstance/syncsetinstance_controller.go:321]: Done, didn't add a unit test in this spot because the harness doesn't seem to like an empty namespace and I don't really want to get into that right now.

**Comment by dgoodwin** [comment]
PR Updated.
